File: runtime/perl/units/wasi/fd_datasync.pl
# requires:
# Core perl exposes no fdatasync; IO::Handle::sync (fsync) is the faithful superset (macOS python does the same).
use IO::Handle ();

sub wasi_fd_datasync {
    my ($self, $fd) = @_;
    my $e = $self->{fds}{$fd};
    return ERRNO_BADF if !defined($e) || $e->{dir};
    return ERRNO_IO unless defined $e->{fh}->sync;
    return ERRNO_SUCCESS;
}

File: crates/dewasm-backend-ruby/units/wasi/fd_datasync.rb
def wasi_fd_datasync(fd)
  io = @fds[fd]
  return ERRNO_BADF if io.nil? || io.is_a?(WasiDir)
  begin
    io.fdatasync
  rescue NotImplementedError, NoMethodError
    io.fsync # not available on all platforms (e.g. macOS) or IO-likes
  end
  ERRNO_SUCCESS
rescue SystemCallError
  ERRNO_IO
end
